## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [78]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data

df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [76]:
# Use the describe function to gather some basic statistics
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  23538.000000  2.353800e+04  23538.000000  23538.000000  23538.000000   
mean   54804.860396  5.483775e+06     11.393831   2022.992693    128.232610   
std    25537.252715  2.600498e+06      7.061066      0.085172    179.429438   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.130000   
25%    33367.000000  3.196372e+06      6.000000   2023.000000     20.990000   
50%    52885.000000  5.535599e+06     11.000000   2023.000000     60.830000   
75%    78618.000000  7.743380e+06     17.000000   2023.000000    169.540000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight            qty   line_number  line_subtotal  \
count  23538.000000  23538.000000   23538.000000  23538.000000   2.353800e+04   
mean      94.409291      5.025105     532.552766      2.990696   7.462013e+04   
std      127.569407      5.524071    9169.006458      2.437117   1.875348e+06   
min        0.100000      0.000000       0.000000      0.000000   0.000000e+00   
25%       14.210000      1.490000      32.000000      1.000000   1.135327e+03   
50%       42.720000      3.280000      68.000000      3.000000   4.355415e+03   
75%      116.950000      6.890000     169.000000      5.000000   1.614738e+04   
max      846.270000     46.430000  842581.000000      9.000000   2.061214e+08   

       total_weight  Shipping_price  
count  2.353800e+04    2.353800e+04  
mean   2.499000e+03    1.750613e+04  
std    5.134985e+04    3.594483e+05  
min    0.000000e+00    0.000000e+00  
25%    6.972000e+01    4.947950e+02  
50%    2.220400e+02    1.554280e+03  
75%    7.188900e+02    5.032230e+03  
max    5.835504e+06    4.084853e+07

In [79]:
# Use this space to do any additional research
# and familiarize yourself with the data.

df.tail()


first       last                      job               phone  \
54634   Lauren      Reese           Radio producer  977.877.5272x11382   
54635  Derrick      Moore                 Musician        358.661.5483   
54636   Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637    Wanda    Solomon             Toxicologist       (311)767-4924   
54638   Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                    email  client_id  order_id  order_date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                    belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5

In [9]:
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [80]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)


category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [48]:
# For the category with the most entries,
# which subcategory had the most entries?
df= df.loc[df["category"] == "consumables"]

df['subcategory'].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [98]:
# Which five clients had the most entries in the data?

df["client_id"].value_counts().head()



client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [117]:
# Store the client ids of those top 5 clients in a list.


top_5_clients = df["client_id"].value_counts().head().index.tolist()

print(top_5_clients)


[33615, 66037, 46820, 38378, 24741]


In [122]:
# How many total units (the qty column) did the
# client with the most entries order order?

df.groupby('client_id')['qty'].sum().max()




4033302

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [150]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty



line_subtotal = df['unit_price']*df['qty']
df["line_subtotal"] = line_subtotal



df[["unit_price","qty", "line_subtotal"]].head()



unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [130]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
total_weight = df['unit_weight']*df['qty']
df["total_weight"] = total_weight
def shipping(total_weight):
    if total_weight < 50:
        return total_weight * 10
    return total_weight * 7

df['Shipping_price'] = df['total_weight'].apply(shipping)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number  line_subtotal  total_weight  \
0     762.71         7.50  105            1      115164.00        787.50   
1      15.09         1.49   21            0         523.95         31.29   
2       7.86         1.68   39            6         527.28         65.52   
3      24.85         1.23   29            3        1056.18         35.67   
4     108.17        46.43   20            1        3902.00        928.60   

   Shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20  

[5 rows x 21 columns]

In [138]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

line_price = (df['Shipping_price'] + df['line_subtotal']) * 1.0925
df["line_price"] = line_price

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number  line_subtotal  total_weight  Shipping_price  \
0         7.50  105            1      115164.00        787.50         5512.50   
1         1.49   21            0         523.95         31.29          312.90   
2         1.68   39            6         527.28         65.52          458.64   
3         1.23   29            3        1056.18         35.67          356.70   
4        46.43   20            1        3902.00        928.60         6500.20   

      line_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 22 columns]

In [143]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).

line_cost = df['unit_cost'] * df['qty'] + df['Shipping_price'] 
df["line_cost"] = line_cost

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number  line_subtotal  total_weight  Shipping_price     line_price  \
0            1      115164.00        787.50         5512.50  131839.076250   
1            0         523.95         31.29          312.90     914.258625   
2            6         527.28         65.52          458.64    1077.117600   
3            3        1056.18         35.67          356.70    1543.571400   
4            1        3902.00        928.60         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 23 columns]

In [148]:
# Create a column for the profit of
# each line using line cost and line price

line_profit =  df['line_price'] - df['line_cost']
df["line_profit"] = line_profit

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   
3          17        2023  ...     24.85        1.23   29            3   
4          21        2023  ...    108.17       46.43   20            1   

   line_subtotal  total_weight  Shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   
3        1056.18         35.67          356.70    1543.571400    1077.35   
4        3902.00        928.60         6500.20   11364.403500    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above


Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.




33615: $8377308.52
66037: $10259514.79
46820: $9743794.36
38378: $12906550.87
24741: $82268892.02


In [17]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.



client_id     qty  shipping_price   line_price    line_cost  line_profit
4      24741  239862      5126448.37  82268892.02  45688899.71  36579992.31
3      38378   73667      3429455.40  12906550.87   9634720.98   3271829.89
1      66037   43018      1395151.85  10259514.79   7004482.98   3255031.81
2      46820   75768      1601448.84   9743794.36   7007191.64   2736602.72
0      33615   64313      1828984.89   8377308.52   6175313.91   2201994.61

In [18]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [19]:
# Sort the updated data by "Total Profit" form highest to lowest


Client ID   Units Shipping Total Revenue Total Cost Total Profit
4      24741  239862   $5.13M       $82.27M    $45.69M      $36.58M
3      38378   73667   $3.43M       $12.91M     $9.63M       $3.27M
1      66037   43018   $1.40M       $10.26M     $7.00M       $3.26M
2      46820   75768   $1.60M        $9.74M     $7.01M       $2.74M
0      33615   64313   $1.83M        $8.38M     $6.18M       $2.20M